### ISSUE TITLE
S4070 Allows non-exclusive values for Flags

### ISSUE THREAD

**Issue by alexangas** [SonarSource/sonar-csharp#1485]
### Description

Currently [rule S4070](https://rules.sonarsource.com/csharp/RSPEC-4070) allows code that breaks the guidelines described in the [Remarks for FlagsAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.flagsattribute?view=netframework-4.7.1#remarks), specifically:

> Define enumeration constants in powers of two, that is, 1, 2, 4, 8, and so on. This means the individual flags in combined enumeration constants do not overlap.

This means that code can be written such that e.g. [Enum.HasFlag()](https://docs.microsoft.com/en-us/dotnet/api/system.enum.hasflag?view=netframework-4.7.1#System_Enum_HasFlag_System_Enum_) does not behave as expected (a value of Yellow also matches Red and Orange), and this is not reported by S4070.

### Repro steps

Given the following code and all rules enabled for SonarC# in the quality profile:

```csharp
[Flags]
public enum Color
{
    None = 0,
    Red = 1,
    Orange = 2,
    Yellow = 3
}
```

### Expected behavior

Expected that S4070 would be triggered for the final value of `Yellow`, as it is not exclusive from `Red` or `Orange`.

### Actual behavior

Neither S4070 or any other SonarC# rule is triggered.

### Known workarounds

None.

### Related information

* SonarC# Version: 7.2 (build 5463)
* Visual Studio Version: 2017
* If running through the Scanner for MSBuild, its version: 3.1.0.1141

**Comment by valhristov** [comment]
S4070 does not raise issue on `Yellow` because its value actually is `Red | Orange`... In other words, the rule raises only when the values are not powers of 2 or are not sums of existing values. A better way to declare the enum would be (I guess you want to enforce this), but that would be another rule:
```
[Flags]
public enum Color
{
    None = 0,
    Red = 1,
    Yellow = 2,
    Orange = Red | Yellow
}
```
I changed the names a bit to ensure color corectness...

**Comment by alexangas** [comment]
I used Color because that's what's in the Sonar code example. Now perhaps I have confused my point! Which is...

As in the MS guidelines linked to above, the point of Flags is to be able to correctly XOR them, and use methods such as Enum.HasFlags with expected and predictable results.

This rule appears to at best ignore this and at worst promote the opposite. Why?

**Comment by Evangelink** [comment]
@valhristov I think that @alexangas is right. When looking at the MSDN article you can see that combination of values is allowed for common use-cases by concatenating the different values. So in this case `Orange` (`Red` OR `Yellow`) should not be allowed but `RedYellow` (maybe we should also allow to have `And` between the words) should be fine.

@alexangas Did you had anything else in mind?

**Comment by valhristov** [comment]
Ok, may be we are talking about different things in the end... :)

As far as I can understand @Evangelink wants to enforce the naming. For example, in my example above, according to you, `Orange` should be named `RedAndYellow` or `RedYellow`, am I right? I could understand the reasoning, but I think it would be difficult to have low false positive rate with a naming convention rule. In addition, if we have such rule I think it should be a separate RSPEC than S4070.

@alexangas are you suggesting to remove the exception for combined values, e.g. `Orange` in my example to be a not allowed value? If this is what you suggest, it would contradict with the third point in the MSDN article, which says that commonly used flag combinations should be allowed.

That's also what I was trying to explain with different words - the current implementation of the rule allows combinations of values, regardless of how they are defined - as an OR expression, or as a number. I think we should prevent the numeric declaration and only allow OR expressions when declaring combinations, for example:
```
Orange = 3 // Noncompliant
Orange = Red | Yellow // Compliant
```

PS: Now when I wrote all of that I think that we should incorporate the logic in the current rule...

**Comment by valhristov** [comment]
I agree, please create the issue for the rule and we will create an [RSPEC](https://jira.sonarsource.com/browse/RSPEC).
